sundials: new test API

AcriusWinter opened the pull request:
Supersedes #35807 (for one package)

Update standalone test API. ~See below comment for output.~

Example installations and associated test results:
```
$ spack find -v sundials
..
sundials@7.1.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda~examples+examples-install~f2003~fcmix~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=14 generator=make logging-level=2 precision=double
sundials@7.1.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda+examples+examples-install~f2003~fcmix~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=14 generator=make logging-level=2 precision=double
==> 2 installed packages
```

```
$ spack -v test run sundials
==> Spack test g46z5xjn2itetpkpyf65tpl6ngpvs4lq
==> Testing package sundials-7.1.1-5nh6oma
==> [2024-08-21-11:52:56.506137] test: test_cvadvdiff_cuda: build and run CUDA cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_cuda: Package must be installed with +cuda+CVODE
==> [2024-08-21-11:52:56.506701] test: test_cvadvdiff_hip: build and run ROCM cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_hip: Package must be installed with +rocm+CVODE
==> [2024-08-21-11:52:56.507035] test: test_cvadvdiff_serial: build and run serial cvAdvDiff_bnd
SKIPPED: Sundials::test_cvadvdiff_serial: Example 'cvAdvDiff_bnd' source directory not found in 7.1.1
==> [2024-08-21-11:52:56.508059] test: test_nvector_cuda: build and run CUDA N_Vector
SKIPPED: Sundials::test_nvector_cuda: Package must be installed with +cuda
==> [2024-08-21-11:52:56.508411] test: test_nvector_hip: build and run ROCM N_Vector
SKIPPED: Sundials::test_nvector_hip: Package must be installed with +rocm
==> [2024-08-21-11:52:56.508697] test: test_nvector_serial: build and run serial N_Vector
SKIPPED: Sundials::test_nvector_serial: Example 'test_nvector_serial' source directory not found in 7.1.1
==> [2024-08-21-11:52:56.509161] test: test_nvector_sycl: build and run SYCL N_Vector
SKIPPED: Sundials::test_nvector_sycl: Package must be installed with +sycl
==> [2024-08-21-11:52:56.509459] test: test_sycl_cvode: build and run SYCL cvAdvDiff_kry
SKIPPED: Sundials::test_sycl_cvode: Package must be installed with +sycl and +CVODE
==> [2024-08-21-11:52:56.511022] Completed testing
==> [2024-08-21-11:52:56.511147] 
======================= SUMMARY: sundials-7.1.1-5nh6oma ========================
Sundials::test_cvadvdiff_cuda .. SKIPPED
Sundials::test_cvadvdiff_hip .. SKIPPED
Sundials::test_cvadvdiff_serial .. SKIPPED
Sundials::test_nvector_cuda .. SKIPPED
Sundials::test_nvector_hip .. SKIPPED
Sundials::test_nvector_serial .. SKIPPED
Sundials::test_nvector_sycl .. SKIPPED
Sundials::test_sycl_cvode .. SKIPPED
============================= 8 skipped of 8 parts =============================
==> Testing package sundials-7.1.1-5srg42r
==> [2024-08-21-11:52:57.763772] test: test_cvadvdiff_cuda: build and run CUDA cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_cuda: Package must be installed with +cuda+CVODE
==> [2024-08-21-11:52:57.764372] test: test_cvadvdiff_hip: build and run ROCM cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_hip: Package must be installed with +rocm+CVODE
==> [2024-08-21-11:52:57.764695] test: test_cvadvdiff_serial: build and run serial cvAdvDiff_bnd
..
==> [2024-08-21-11:53:03.501391] '/usr/bin/make'
[  3%] Building C object CMakeFiles/cvAdvDiff_bnd.dir/cvAdvDiff_bnd.c.o
[  6%] Linking C executable cvAdvDiff_bnd
[  6%] Built target cvAdvDiff_bnd
[ 10%] Building C object CMakeFiles/cvAnalytic_mels.dir/cvAnalytic_mels.c.o
[ 13%] Linking C executable cvAnalytic_mels
[ 13%] Built target cvAnalytic_mels
[ 16%] Building C object CMakeFiles/cvDirectDemo_ls.dir/cvDirectDemo_ls.c.o
[ 20%] Linking C executable cvDirectDemo_ls
[ 20%] Built target cvDirectDemo_ls
[ 23%] Building C object CMakeFiles/cvDisc_dns.dir/cvDisc_dns.c.o
[ 26%] Linking C executable cvDisc_dns
[ 26%] Built target cvDisc_dns
[ 30%] Building C object CMakeFiles/cvDiurnal_kry_bp.dir/cvDiurnal_kry_bp.c.o
[ 33%] Linking C executable cvDiurnal_kry_bp
[ 33%] Built target cvDiurnal_kry_bp
[ 36%] Building C object CMakeFiles/cvDiurnal_kry.dir/cvDiurnal_kry.c.o
[ 40%] Linking C executable cvDiurnal_kry
[ 40%] Built target cvDiurnal_kry
[ 43%] Building C object CMakeFiles/cvKrylovDemo_ls.dir/cvKrylovDemo_ls.c.o
[ 46%] Linking C executable cvKrylovDemo_ls
[ 46%] Built target cvKrylovDemo_ls
[ 50%] Building C object CMakeFiles/cvKrylovDemo_prec.dir/cvKrylovDemo_prec.c.o
[ 53%] Linking C executable cvKrylovDemo_prec
[ 53%] Built target cvKrylovDemo_prec
[ 56%] Building C object CMakeFiles/cvParticle_dns.dir/cvParticle_dns.c.o
[ 60%] Linking C executable cvParticle_dns
[ 60%] Built target cvParticle_dns
[ 63%] Building C object CMakeFiles/cvPendulum_dns.dir/cvPendulum_dns.c.o
[ 66%] Linking C executable cvPendulum_dns
[ 66%] Built target cvPendulum_dns
[ 70%] Building C object CMakeFiles/cvRoberts_dns.dir/cvRoberts_dns.c.o
[ 73%] Linking C executable cvRoberts_dns
[ 73%] Built target cvRoberts_dns
[ 76%] Building C object CMakeFiles/cvRoberts_dns_constraints.dir/cvRoberts_dns_constraints.c.o
[ 80%] Linking C executable cvRoberts_dns_constraints
[ 80%] Built target cvRoberts_dns_constraints
[ 83%] Building C object CMakeFiles/cvRoberts_dns_negsol.dir/cvRoberts_dns_negsol.c.o
[ 86%] Linking C executable cvRoberts_dns_negsol
[ 86%] Built target cvRoberts_dns_negsol
[ 90%] Building C object CMakeFiles/cvRoberts_dns_uw.dir/cvRoberts_dns_uw.c.o
[ 93%] Linking C executable cvRoberts_dns_uw
[ 93%] Built target cvRoberts_dns_uw
[ 96%] Building C object CMakeFiles/cvRocket_dns.dir/cvRocket_dns.c.o
[100%] Linking C executable cvRocket_dns
[100%] Built target cvRocket_dns
==> [2024-08-21-11:53:10.601388] 'cvAdvDiff_bnd'

2-D Advection-Diffusion Equation
Mesh dimensions = 10 X 5
Total system size = 50
Tolerance parameters: reltol = 0   abstol = 1e-05

At t = 0      max.norm(u) =  8.954716e+01 
At t = 0.10   max.norm(u) =  4.132889e+00   nst =   85
At t = 0.20   max.norm(u) =  1.039294e+00   nst =  103
At t = 0.30   max.norm(u) =  2.979829e-01   nst =  113
At t = 0.40   max.norm(u) =  8.765774e-02   nst =  120
At t = 0.50   max.norm(u) =  2.625637e-02   nst =  126
At t = 0.60   max.norm(u) =  7.830425e-03   nst =  130
At t = 0.70   max.norm(u) =  2.329387e-03   nst =  134
At t = 0.80   max.norm(u) =  6.953434e-04   nst =  137
At t = 0.90   max.norm(u) =  2.115983e-04   nst =  140
At t = 1.00   max.norm(u) =  6.556853e-05   nst =  142

Final Statistics:
nst = 142    nfe  = 173    nsetups = 23     nfeLS = 0      nje = 3
nni = 170    ncfn = 0      netf = 3
==> [2024-08-21-11:53:10.716146] '/usr/bin/make' 'clean'
PASSED: Sundials::test_cvadvdiff_serial
==> [2024-08-21-11:53:12.228656] test: test_nvector_cuda: build and run CUDA N_Vector
SKIPPED: Sundials::test_nvector_cuda: Package must be installed with +cuda
==> [2024-08-21-11:53:12.229396] test: test_nvector_hip: build and run ROCM N_Vector
SKIPPED: Sundials::test_nvector_hip: Package must be installed with +rocm
==> [2024-08-21-11:53:12.229772] test: test_nvector_serial: build and run serial N_Vector
..
==> [2024-08-21-11:53:12.378480] '/usr/bin/make'
..
In file included from $spack/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/sundials-7.1.1-5srg42ra2etlprjjagagncnezxxmmck5/include/sundials/sundials_math.h:26,
                 from test_nvector.c:33:
$spack/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/sundials-7.1.1-5srg42ra2etlprjjagagncnezxxmmck5/include/sundials/sundials_types.h:57:10: fatal error: mpi.h: No such file or directory
   57 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
make: *** [Makefile:90: test_nvector.o] Error 1
FAILED: Sundials::test_nvector_serial: Command exited with status 2:
..
==> [2024-08-21-11:53:12.485642] test: test_nvector_sycl: build and run SYCL N_Vector
SKIPPED: Sundials::test_nvector_sycl: Package must be installed with +sycl
==> [2024-08-21-11:53:12.486049] test: test_sycl_cvode: build and run SYCL cvAdvDiff_kry
SKIPPED: Sundials::test_sycl_cvode: Package must be installed with +sycl and +CVODE
==> [2024-08-21-11:53:12.487596] Completed testing
==> [2024-08-21-11:53:12.487729] 
======================= SUMMARY: sundials-7.1.1-5srg42r ========================
Sundials::test_cvadvdiff_cuda .. SKIPPED
Sundials::test_cvadvdiff_hip .. SKIPPED
Sundials::test_cvadvdiff_serial .. PASSED
Sundials::test_nvector_cuda .. SKIPPED
Sundials::test_nvector_hip .. SKIPPED
Sundials::test_nvector_serial .. FAILED
Sundials::test_nvector_sycl .. SKIPPED
Sundials::test_sycl_cvode .. SKIPPED
=================== 6 skipped, 1 passed, 1 failed of 8 parts ===================
==> [2024-08-21-11:53:12.487796] 
..
======================== 1 skipped, 1 failed of 2 specs ========================
```

AcriusWinter committed:
sundials: new test API

AcriusWinter commented:
======================= SUMMARY: sundials-7.0.0-ifkzwot ========================
Sundials::test_install .. PASSED
Sundials::test_part_('nvector/serial/test_nvector_serial', ['10', '0'], 'Test serial N_Vector', False) .. SKIPPED
Sundials::test_part_('cvode/serial/cvAdvDiff_bnd', [], 'Test CVODE', True) .. PASSED
Sundials::test_smoke .. PASSED
======================== 3 passed, 1 skipped of 4 parts ========================
============================== 1 passed of 1 spec ==============================

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -832,31 +832,35 @@ def build_smoke_tests(self):
-                skip_missing=True,
-                purpose=smoke_test[2],
-            )
+        cmake_bin = self.cmake_bin(set=False)
+
+        if not cmake_bin:
+            raise SkipTest("cmake_bin_path.txt not found")
```
This is the `cmake` hack that is supposed to be removed along with lines 795-810.  It follows the same pattern as the other PRs that used it.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -832,31 +832,35 @@ def build_smoke_tests(self):
+                    raise SkipTest(f"{smoke_test[0]} not found in {self.version}")
+                exe(*smoke_test[1])
+
+        make = which("make")
```
This duplicates the assignment at 847.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -832,31 +832,35 @@ def build_smoke_tests(self):
-                self.run_test(exe="make", options=["clean"])
+                if smoke_test[3]:  # use cmake
+                    cmake(".")
+                    make()
 
-    def test(self):
-        self.build_smoke_tests()
-        self.run_smoke_tests()
-        self.clean_smoke_tests()
-        return
+        for smoke_test in self._smoke_tests:
+            with test_part(self, f"test_part_{smoke_test}", purpose=smoke_test[2]):
+                exe = which(join_path(self._smoke_tests_path, smoke_test[0]))
+                if exe is None:
+                    raise SkipTest(f"{smoke_test[0]} not found in {self.version}")
+                exe(*smoke_test[1])
+
+        make = which("make")
+        work_dir = join_path(self._smoke_tests_path, os.path.dirname(smoke_test[0]))
+        with working_dir(work_dir):
+            for smoke_test in self._smoke_tests:
+                make("clean")
```
Looks to me like you only need to iterate over the smoke tests once, use one `working_dir()` context manager, and do the test-specific `cmake`, `make`, run, and `make(clean)` once per test.

Unless there is a good reason not to, there should be a separate test methods for each of the tests from `_smoke_tests` (so that method can go away).

tldahlgren replied:
You also want to make sure you install the software s.t. you can run all of the tests.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -832,31 +832,35 @@ def build_smoke_tests(self):
+        cmake = which(cmake_bin)
+        make = which("make")
         for smoke_test in self._smoke_tests:
             work_dir = join_path(self._smoke_tests_path, os.path.dirname(smoke_test[0]))
             with working_dir(work_dir):
-                self.run_test(exe="make", options=["clean"])
+                if smoke_test[3]:  # use cmake
+                    cmake(".")
+                    make()
```
Looks like this is intended to replace lines 817-833, which should then be removed.

tldahlgren reviewed (changes requested):
See comments.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -832,31 +832,35 @@ def build_smoke_tests(self):
-        self.clean_smoke_tests()
-        return
+        for smoke_test in self._smoke_tests:
+            with test_part(self, f"test_part_{smoke_test}", purpose=smoke_test[2]):
```
The test name needs to be a string that does not contain spaces, commas, slashes, brackets, parentheses, or quotes.

AcriusWinter committed:
refactor

AcriusWinter commented:
======================= SUMMARY: sundials-7.1.1-jr7vkf2 ========================
Sundials::test_cuda .. SKIPPED
Sundials::test_cvode .. PASSED
Sundials::test_install .. PASSED
Sundials::test_cuda_basic .. FAILED
Sundials::test_cuda_CVODE_HIP .. FAILED
Sundials::test_rocm .. PASSED
Sundials::test_serial_nvector .. SKIPPED
Sundials::test_sycl .. SKIPPED
=================== 3 skipped, 3 passed, 2 failed of 8 parts ===================
==> [2024-07-24-15:11:18.538191] 

See test results at:
  /g/g16/lawrence31/.spack/test/plyji4nq6esk4mtmv2spsloipjbqdnim/sundials-7.1.1-jr7vkf2-test-out.txt
==> Error: TestFailure: 2 tests failed.


Command exited with status 1:
    '/usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/cmake-3.29.4-eplcyjvolizfqs4g7zvmma6hrfkif4cg/bin/cmake' '.'


1 error found in test log:
     101      /usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hip-6.1.1-kkdbijj77caoyagaxo7fzgxkz
            ztoehit/lib/cmake/hip/hip-config-amd.cmake:99 (find_dependency)
     102      /usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hip-6.1.1-kkdbijj77caoyagaxo7fzgxkz
            ztoehit/lib/cmake/hip/hip-config.cmake:149 (include)
     103      CMakeLists.txt:57 (find_package)
     104    
     105    
     106    -- Configuring incomplete, errors occurred!
  >> 107    FAILED: Sundials::test_cuda_basic: Command exited with status 1:
     108        '/usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/cmake-3.29.4-eplcyjvolizfqs4g7zv
            mma6hrfkif4cg/bin/cmake' '.'
     109      File "/usr/WS2/lawrence31/spack/bin/spack", line 52, in <module>
     110        sys.exit(main())
     111      File "/usr/WS2/lawrence31/spack/lib/spack/spack_installable/main.py", line 42, in main
     112        sys.exit(spack.main.main(argv))
     113      File "/usr/WS2/lawrence31/spack/lib/spack/spack/main.py", line 1070, in main



Command exited with status 1:
    '/usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/cmake-3.29.4-eplcyjvolizfqs4g7zvmma6hrfkif4cg/bin/cmake' '.'


1 error found in test log:
     174      /usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hip-6.1.1-kkdbijj77caoyagaxo7fzgxkz
            ztoehit/lib/cmake/hip/hip-config-amd.cmake:99 (find_dependency)
     175      /usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hip-6.1.1-kkdbijj77caoyagaxo7fzgxkz
            ztoehit/lib/cmake/hip/hip-config.cmake:149 (include)
     176      CMakeLists.txt:57 (find_package)
     177    
     178    
     179    -- Configuring incomplete, errors occurred!
  >> 180    FAILED: Sundials::test_cuda_CVODE_HIP: Command exited with status 1:
     181        '/usr/WS2/lawrence31/spack/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/cmake-3.29.4-eplcyjvolizfqs4g7zv
            mma6hrfkif4cg/bin/cmake' '.'
     182      File "/usr/WS2/lawrence31/spack/bin/spack", line 52, in <module>
     183        sys.exit(main())
     184      File "/usr/WS2/lawrence31/spack/lib/spack/spack_installable/main.py", line 42, in main
     185        sys.exit(spack.main.main(argv))
     186      File "/usr/WS2/lawrence31/spack/lib/spack/spack/main.py", line 1070, in main


See test log for details:
  /g/g16/lawrence31/.spack/test/plyji4nq6esk4mtmv2spsloipjbqdnim/sundials-7.1.1-jr7vkf2-test-out.txt

======================== 1 passed, 1 failed of 2 specs =========================

AcriusWinter committed:
flake8

AcriusWinter committed:
refactor again

AcriusWinter committed:
get rid of self.install_test_root

AcriusWinter committed:
no_self.run_test

AcriusWinter committed:
options

AcriusWinter committed:
correct_mistake

AcriusWinter committed:
namechange

AcriusWinter commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,86 @@ def build_smoke_tests(self):
-    def run_smoke_tests(self):
-        if "~examples-install" in self.spec:
-            return
+    def test_serial_nvector(self):
```
```suggestion
    def test_nvector_serial(self):
```

AcriusWinter committed:
Update var/spack/repos/builtin/packages/sundials/package.py

AcriusWinter commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,86 @@ def build_smoke_tests(self):
+                if not cmake:
+                    raise SkipTest("cmake not found")
+                cmake(".")
+            make()
```
```suggestion
            make = which("make")
            make()
```

AcriusWinter committed:
Update var/spack/repos/builtin/packages/sundials/package.py

AcriusWinter commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,87 @@ def build_smoke_tests(self):
+        with working_dir(work_dir):
+            if cmake_bool:
+                cmake = self.spec["cmake"].command
+                if not cmake:
+                    raise SkipTest("cmake not found")
```
```suggestion
```

AcriusWinter committed:
Update var/spack/repos/builtin/packages/sundials/package.py

AcriusWinter commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,85 @@ def build_smoke_tests(self):
+        if "+sycl" not in self.spec or "+CVODE" not in self.spec:
+            raise SkipTest("Package must be installed with +sycl and +CVODE")
+
+        self.run_sundials("cvode/sycl/cvAdvDiff_kry_sycl", [], True)
```
```suggestion
        self.run_sundials(join_path("cvode", "sycl", "cvAdvDiff_kry_sycl"), [], True)
```

AcriusWinter committed:
Update var/spack/repos/builtin/packages/sundials/package.py

AcriusWinter committed:
fstring

AcriusWinter commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -744,82 +744,22 @@ def libs(self):
     @run_after("install")
     @on_package_attributes(run_tests=True)
-    def test_install(self):
+    def make_test_install(self):
```
```suggestion
    def test_install(self):
```

AcriusWinter committed:
Update var/spack/repos/builtin/packages/sundials/package.py

AcriusWinter committed:
joinPath

AcriusWinter commented:
Please lmk if there are any other changes that need to be made.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -749,77 +749,17 @@ def test_install(self):
-        self.cmake_bin(set=True)
+            install_tree(self._smoke_tests_path, join_path(install_test_root(self), "testing"))
 
     def build_smoke_tests(self):
```
This method is no longer called and uses a property (`self._smoke_tests`) no longer being used.  So it should be removed.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,91 @@ def build_smoke_tests(self):
+
+        self.run_sundials(join_path("cvode", "sycl", "cvAdvDiff_kry_sycl"), [], True)
+
+    def run_sundials(self, exe_path, opts, cmake_bool):
```
This method should appear before the tests as the equivalent had been in the original code.  There are enough changes in the area that it might not reduce the diffs but it has a better chance than it has with this method being moved to the end.

tldahlgren replied:
To be more accurate, this is building and running an example *for* the sundials package.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,91 @@ def build_smoke_tests(self):
-    def clean_smoke_tests(self):
-        if "~examples-install" in self.spec:
-            return
+    def test_cvode(self):
```
This method name isn't very descriptive, especially since there are multiple `cvode` tests.  It should better reflect the actual test being executed.

This is true for most, if not all, of the stand-alone test methods.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,91 @@ def build_smoke_tests(self):
-        if "~examples-install" in self.spec:
-            return
+    def test_cvode(self):
+        """Run CVODE test"""
```
Docstring could also be more descriptive for the same reason OR use the original purpose IF it is more descriptive.

This is true for most, if not all, of the stand-alone test methods.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,91 @@ def build_smoke_tests(self):
+            raise SkipTest("Package must be installed with +sycl")
+
+        self.run_sundials(
+            join_path("nvector", "sycl", "test_nvector_sycl"), ["10", "0", "0"], True
```
```suggestion
            join_path("nvector", "sycl", "test_nvector_sycl"), ["10", "0", "0"], False
```

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -828,35 +768,91 @@ def build_smoke_tests(self):
+        if "~examples-install" in self.spec:
+            raise SkipTest("Package must be installed with +examples-install")
+
+        work_dir = join_path(self._smoke_tests_path, os.path.dirname(exe_path))
```
This is still running under the installation directory (see value of `self._smoke_tests_path`).

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -749,77 +749,17 @@ def test_install(self):
         if "+examples-install" in self.spec:
-            install_tree(self._smoke_tests_path, join_path(self.install_test_root, "testing"))
-        self.cmake_bin(set=True)
+            install_tree(self._smoke_tests_path, join_path(install_test_root(self), "testing"))
```
This is copying all of the installed examples to the test cache, thereby unnecessarily duplicating them.

The preferred approach in this case is to explicitly copy the files to the test stage before using them.

tldahlgren commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -749,77 +749,17 @@ def test_install(self):
         with working_dir(self.build_directory):
```
This property is no longer available so must be `self.builder.build_directory`.  Since it relies on the build stage, it cannot be re-used as a stand-alone test so must be renamed (e.g., `check_test_install`).

tldahlgren committed:
sundials: assorted cleanup
- Makefike -> Makefile
- removed setup_smoke_tests
  * duplicated "installed" files AND wasn't being used
- test_install -> check_test_install
  * since can't run as stand-alone test
- self.build_directory -> self.builder.build_directory
  * build-time tests still fail
- moved run_sundials back to above test methods
- run_sundials -> run_example (ore descriptive)
- skip example if directory doesn't exist (is that sufficient though?)
- changed test method names to be more descriptive of the example
- changed test method docstrings to be more desriptive of the example

tldahlgren reviewed (changes requested):
See comments

tldahlgren reviewed (dismissed):
IMO This PR accomplishes the goal of converting testing from the old API to the new one.

@balos1 @carsonwoods @gardner48 Would appreciate your feedback on how well this works on systems of interest.

tldahlgren committed:
sundials: removed unused import

becker33 commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -744,119 +742,95 @@ def libs(self):
+            join_path("nvector", "sycl", "test_nvector_sycl"), ["10", "0", "0"], False
+        )
+
+    def test_sycl_cvode(self):
+        """build and run SYCL cvAdvDiff_kry"""
+        if "+sycl" not in self.spec or "+CVODE" not in self.spec:
+            raise SkipTest("Package must be installed with +sycl and +CVODE")
+
+        self.run_example(join_path("cvode", "sycl", "cvAdvDiff_kry_sycl"), [], True)
```
Not for this PR necessarily, but would it be possible to decorate these with `@when` instead of having to have the conditional `SkipTest` error?

tldahlgren replied:
Unless something has changed not involving https://github.com/spack/spack/pull/27789, that feature is not supported.  BUT I haven't tested it in a while.

becker33 replied:
Ugh I forgot about that PR, sorry for the noise.

Yeah that would be a great addition later.

becker33 commented on var/spack/repos/builtin/packages/sundials/package.py:
```diff
@@ -744,119 +742,95 @@ def libs(self):
+        (dirname, basename) = os.path.split(exe_path)
+        srcpath = join_path(self._smoke_tests_path, dirname)
+        if not os.path.exists(srcpath):
+            raise SkipTest(f"Example '{dirname}' not found in {self.version}")
```
Shouldn't this be `basename` instead of `dirname`?

tldahlgren replied:
I debated on that one.

If you look at the original and current entries, `basename` is the name of the executable that hasn't been built at this point.  I was trying to avoid the overhead of the context, `Executable`s instantiation, etc. if the source directory doesn't exist.

So what is more informative at this point?

becker33 replied:
I think the basenames carry much more information than dirnames do.

tldahlgren replied:
Done

tldahlgren replied:
FYI. The package installed with `~examples` shows the resulting message for each of the tests.

tldahlgren commented:
> Supersedes #35807 (for one package)
> 
> Update standalone test API. ~See below comment for output.~

Latest test results for selected installations:
```
$ spack find -v sundials
..
sundials@7.1.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda~examples+examples-install~f2003~fcmix~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=14 generator=make logging-level=2 precision=double
sundials@7.1.1+ARKODE+CVODE+CVODES+IDA+IDAS+KINSOL~cuda+examples+examples-install~f2003~fcmix~ginkgo~hypre~int64~ipo~klu~kokkos~kokkos-kernels~lapack~magma~monitoring+mpi~openmp~petsc~profiling~pthread~raja~rocm+shared+static~superlu-dist~superlu-mt~sycl~trilinos build_system=cmake build_type=Release cstd=99 cxxstd=14 generator=make logging-level=2 precision=double
==> 2 installed packages
```

```
$ spack -v test run sundials
==> Spack test g46z5xjn2itetpkpyf65tpl6ngpvs4lq
==> Testing package sundials-7.1.1-5nh6oma
==> [2024-08-21-11:52:56.506137] test: test_cvadvdiff_cuda: build and run CUDA cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_cuda: Package must be installed with +cuda+CVODE
==> [2024-08-21-11:52:56.506701] test: test_cvadvdiff_hip: build and run ROCM cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_hip: Package must be installed with +rocm+CVODE
==> [2024-08-21-11:52:56.507035] test: test_cvadvdiff_serial: build and run serial cvAdvDiff_bnd
SKIPPED: Sundials::test_cvadvdiff_serial: Example 'cvAdvDiff_bnd' source directory not found in 7.1.1
==> [2024-08-21-11:52:56.508059] test: test_nvector_cuda: build and run CUDA N_Vector
SKIPPED: Sundials::test_nvector_cuda: Package must be installed with +cuda
==> [2024-08-21-11:52:56.508411] test: test_nvector_hip: build and run ROCM N_Vector
SKIPPED: Sundials::test_nvector_hip: Package must be installed with +rocm
==> [2024-08-21-11:52:56.508697] test: test_nvector_serial: build and run serial N_Vector
SKIPPED: Sundials::test_nvector_serial: Example 'test_nvector_serial' source directory not found in 7.1.1
==> [2024-08-21-11:52:56.509161] test: test_nvector_sycl: build and run SYCL N_Vector
SKIPPED: Sundials::test_nvector_sycl: Package must be installed with +sycl
==> [2024-08-21-11:52:56.509459] test: test_sycl_cvode: build and run SYCL cvAdvDiff_kry
SKIPPED: Sundials::test_sycl_cvode: Package must be installed with +sycl and +CVODE
==> [2024-08-21-11:52:56.511022] Completed testing
==> [2024-08-21-11:52:56.511147] 
======================= SUMMARY: sundials-7.1.1-5nh6oma ========================
Sundials::test_cvadvdiff_cuda .. SKIPPED
Sundials::test_cvadvdiff_hip .. SKIPPED
Sundials::test_cvadvdiff_serial .. SKIPPED
Sundials::test_nvector_cuda .. SKIPPED
Sundials::test_nvector_hip .. SKIPPED
Sundials::test_nvector_serial .. SKIPPED
Sundials::test_nvector_sycl .. SKIPPED
Sundials::test_sycl_cvode .. SKIPPED
============================= 8 skipped of 8 parts =============================
==> Testing package sundials-7.1.1-5srg42r
==> [2024-08-21-11:52:57.763772] test: test_cvadvdiff_cuda: build and run CUDA cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_cuda: Package must be installed with +cuda+CVODE
==> [2024-08-21-11:52:57.764372] test: test_cvadvdiff_hip: build and run ROCM cvAdvDiff_kry
SKIPPED: Sundials::test_cvadvdiff_hip: Package must be installed with +rocm+CVODE
==> [2024-08-21-11:52:57.764695] test: test_cvadvdiff_serial: build and run serial cvAdvDiff_bnd
..
==> [2024-08-21-11:53:03.501391] '/usr/bin/make'
[  3%] Building C object CMakeFiles/cvAdvDiff_bnd.dir/cvAdvDiff_bnd.c.o
[  6%] Linking C executable cvAdvDiff_bnd
[  6%] Built target cvAdvDiff_bnd
[ 10%] Building C object CMakeFiles/cvAnalytic_mels.dir/cvAnalytic_mels.c.o
[ 13%] Linking C executable cvAnalytic_mels
[ 13%] Built target cvAnalytic_mels
[ 16%] Building C object CMakeFiles/cvDirectDemo_ls.dir/cvDirectDemo_ls.c.o
[ 20%] Linking C executable cvDirectDemo_ls
[ 20%] Built target cvDirectDemo_ls
[ 23%] Building C object CMakeFiles/cvDisc_dns.dir/cvDisc_dns.c.o
[ 26%] Linking C executable cvDisc_dns
[ 26%] Built target cvDisc_dns
[ 30%] Building C object CMakeFiles/cvDiurnal_kry_bp.dir/cvDiurnal_kry_bp.c.o
[ 33%] Linking C executable cvDiurnal_kry_bp
[ 33%] Built target cvDiurnal_kry_bp
[ 36%] Building C object CMakeFiles/cvDiurnal_kry.dir/cvDiurnal_kry.c.o
[ 40%] Linking C executable cvDiurnal_kry
[ 40%] Built target cvDiurnal_kry
[ 43%] Building C object CMakeFiles/cvKrylovDemo_ls.dir/cvKrylovDemo_ls.c.o
[ 46%] Linking C executable cvKrylovDemo_ls
[ 46%] Built target cvKrylovDemo_ls
[ 50%] Building C object CMakeFiles/cvKrylovDemo_prec.dir/cvKrylovDemo_prec.c.o
[ 53%] Linking C executable cvKrylovDemo_prec
[ 53%] Built target cvKrylovDemo_prec
[ 56%] Building C object CMakeFiles/cvParticle_dns.dir/cvParticle_dns.c.o
[ 60%] Linking C executable cvParticle_dns
[ 60%] Built target cvParticle_dns
[ 63%] Building C object CMakeFiles/cvPendulum_dns.dir/cvPendulum_dns.c.o
[ 66%] Linking C executable cvPendulum_dns
[ 66%] Built target cvPendulum_dns
[ 70%] Building C object CMakeFiles/cvRoberts_dns.dir/cvRoberts_dns.c.o
[ 73%] Linking C executable cvRoberts_dns
[ 73%] Built target cvRoberts_dns
[ 76%] Building C object CMakeFiles/cvRoberts_dns_constraints.dir/cvRoberts_dns_constraints.c.o
[ 80%] Linking C executable cvRoberts_dns_constraints
[ 80%] Built target cvRoberts_dns_constraints
[ 83%] Building C object CMakeFiles/cvRoberts_dns_negsol.dir/cvRoberts_dns_negsol.c.o
[ 86%] Linking C executable cvRoberts_dns_negsol
[ 86%] Built target cvRoberts_dns_negsol
[ 90%] Building C object CMakeFiles/cvRoberts_dns_uw.dir/cvRoberts_dns_uw.c.o
[ 93%] Linking C executable cvRoberts_dns_uw
[ 93%] Built target cvRoberts_dns_uw
[ 96%] Building C object CMakeFiles/cvRocket_dns.dir/cvRocket_dns.c.o
[100%] Linking C executable cvRocket_dns
[100%] Built target cvRocket_dns
==> [2024-08-21-11:53:10.601388] 'cvAdvDiff_bnd'

2-D Advection-Diffusion Equation
Mesh dimensions = 10 X 5
Total system size = 50
Tolerance parameters: reltol = 0   abstol = 1e-05

At t = 0      max.norm(u) =  8.954716e+01 
At t = 0.10   max.norm(u) =  4.132889e+00   nst =   85
At t = 0.20   max.norm(u) =  1.039294e+00   nst =  103
At t = 0.30   max.norm(u) =  2.979829e-01   nst =  113
At t = 0.40   max.norm(u) =  8.765774e-02   nst =  120
At t = 0.50   max.norm(u) =  2.625637e-02   nst =  126
At t = 0.60   max.norm(u) =  7.830425e-03   nst =  130
At t = 0.70   max.norm(u) =  2.329387e-03   nst =  134
At t = 0.80   max.norm(u) =  6.953434e-04   nst =  137
At t = 0.90   max.norm(u) =  2.115983e-04   nst =  140
At t = 1.00   max.norm(u) =  6.556853e-05   nst =  142

Final Statistics:
nst = 142    nfe  = 173    nsetups = 23     nfeLS = 0      nje = 3
nni = 170    ncfn = 0      netf = 3
==> [2024-08-21-11:53:10.716146] '/usr/bin/make' 'clean'
PASSED: Sundials::test_cvadvdiff_serial
==> [2024-08-21-11:53:12.228656] test: test_nvector_cuda: build and run CUDA N_Vector
SKIPPED: Sundials::test_nvector_cuda: Package must be installed with +cuda
==> [2024-08-21-11:53:12.229396] test: test_nvector_hip: build and run ROCM N_Vector
SKIPPED: Sundials::test_nvector_hip: Package must be installed with +rocm
==> [2024-08-21-11:53:12.229772] test: test_nvector_serial: build and run serial N_Vector
..
==> [2024-08-21-11:53:12.378480] '/usr/bin/make'
..
In file included from $spack/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/sundials-7.1.1-5srg42ra2etlprjjagagncnezxxmmck5/include/sundials/sundials_math.h:26,
                 from test_nvector.c:33:
$spack/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/sundials-7.1.1-5srg42ra2etlprjjagagncnezxxmmck5/include/sundials/sundials_types.h:57:10: fatal error: mpi.h: No such file or directory
   57 | #include <mpi.h>
      |          ^~~~~~~
compilation terminated.
make: *** [Makefile:90: test_nvector.o] Error 1
FAILED: Sundials::test_nvector_serial: Command exited with status 2:
..
==> [2024-08-21-11:53:12.485642] test: test_nvector_sycl: build and run SYCL N_Vector
SKIPPED: Sundials::test_nvector_sycl: Package must be installed with +sycl
==> [2024-08-21-11:53:12.486049] test: test_sycl_cvode: build and run SYCL cvAdvDiff_kry
SKIPPED: Sundials::test_sycl_cvode: Package must be installed with +sycl and +CVODE
==> [2024-08-21-11:53:12.487596] Completed testing
==> [2024-08-21-11:53:12.487729] 
======================= SUMMARY: sundials-7.1.1-5srg42r ========================
Sundials::test_cvadvdiff_cuda .. SKIPPED
Sundials::test_cvadvdiff_hip .. SKIPPED
Sundials::test_cvadvdiff_serial .. PASSED
Sundials::test_nvector_cuda .. SKIPPED
Sundials::test_nvector_hip .. SKIPPED
Sundials::test_nvector_serial .. FAILED
Sundials::test_nvector_sycl .. SKIPPED
Sundials::test_sycl_cvode .. SKIPPED
=================== 6 skipped, 1 passed, 1 failed of 8 parts ===================
==> [2024-08-21-11:53:12.487796] 
..
======================== 1 skipped, 1 failed of 2 specs ========================
```

tldahlgren committed:
sundials: use test basename in message (per feedback)

tldahlgren reviewed (approved):
Accomplishes the goal of using the new test API even if there are issues with building/running the examples.

tldahlgren commented:
@becker33 Are you okay with this being merged?
